### PR TITLE
[spec] Document `noreturn`

### PR DESCRIPTION
**Pull request by ntrel**
See also: https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1034.md

### CODE REVIEW

**Comment by maxhaton** [comment]
@dkorpel

**Comment by ntrel** [comment]
As documented in the DIP, I tried a throwing lambda, function pointers with other return types accepting a noreturn function, and template inference for `[]` to `T[]` and `null` to `T*` but none of these things work yet with stable dmd. Shall I file enhancements for these?

BTW it would be good to merge this pull to stable as well once it is approved.

**Comment by ntrel** [comment]
> it would be good to merge this pull to stable 

I have split out the changes to #3150 for `main` return type because that isn't stable yet.

**Comment by MoonlightSentinel** [comment]
> As documented in the DIP, I tried a throwing lambda, function pointers with other return types accepting a noreturn function, and template inference for `[]` to `T[]` and `null` to `T*` but none of these things work yet with stable dmd. Shall I file enhancements for these?

Sure. Note that there's already a [draft PR for `throw` expressions](dlang/dmd#13162) .

> BTW it would be good to merge this pull to stable as well once it is approved.

This will happen anyways with the regular master - stable merge and there is no rush given the state of `noreturn` (very WIP).

File: spec/type.dd
Comment (MoonlightSentinel) [inline spec/type.dd:586]: This / another example should probably include usages of `noreturn` outside of function return types, i.e. variables.

Comment (ntrel) [reply spec/type.dd:586]: @MoonlightSentinel Do you have an idea for an example? The DIP only seems to mention `noreturn[] storage;` for `MapResult`, and stable dmd doesn't seem to have the enhanced compile-time understanding of this (which might make a good example).

Comment (MoonlightSentinel) [reply spec/type.dd:586]: Unlike void, `noreturn` can be used everywhere a "normal" type may appear, i.e. variables, members, parameters, ... . 

Maybe add an example where the "return value" of a `noreturn` variable is stored in a variable?

**Comment by ntrel** [comment]
@MoonlightSentinel 
> This will happen anyways with the regular master - stable merge

Ok.

> there is no rush given the state of noreturn (very WIP).

It would be good to merge this and the dependent `main` PR before the next release. It seems what I have documented is quite stable. I haven't added a variable example because I don't know if this is stable enough for a meaningful example.

File: spec/type.dd
Comment (dkorpel) [inline spec/type.dd:570]: I don't think sentence adds value, it's unspecific.

Comment (ntrel) [reply spec/type.dd]: Removed now.

**Review by dkorpel — APPROVED** [review]

**Pull request merged**
